torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -0,0 +1,109 @@
+import math
+
+import numpy as np
+import torch
+import torch.distributed as dist
+
+
+def _orthogonalize(matrix, epsilon=1e-8):
+    """Applies Gram-Schmidt procedure to orthogonalize a given 2D tensor."""
+    num_cols = matrix.shape[1]
+    for i in range(num_cols):
+        # Normalize the i'th column.
+        col = matrix[:, i : i + 1]
+        # Note that col ** 2 can underflow/overflow if we use FP16.
+        # May need to consder multiplying a scaling factor and divding it later, or using bfloat16 isntead.
+        col /= torch.sqrt(torch.sum(col ** 2))
+        # If no epsilon is added here, division by zero may be caused by vanishing gradients.
+        # This epsilon is not needed if the input matrix covers the gradients of at least one entire layer in the neural network.
+        if epsilon > 0:
+            col += epsilon
+        # Project it on the rest and remove it.
+        if i + 1 < num_cols:
+            rest = matrix[:, i + 1 :]
+            rest -= torch.sum(col * rest, dim=0) * col
+
+
+def _set_random(vector, random_seed=0):
+    """
+    Initializes a 1D tensor from a standard normal distribution.
+    The see makes sure that the randomized vector in all the DDP replicas are the same at every step.
+    """
+    torch.manual_seed(np.random.RandomState(random_seed).randint(1_000_000_000))
+    vector.data[:] = torch.randn(*vector.shape)
+
+
+def powerSGD_hook(
+    process_group: object, bucket: dist._GradBucket, rank: int = 1
+) -> torch.futures.Future:
+    """
+    This DDP communication hook implements a simplified PowerSGD graidient compression
+    algorithm described in https://arxiv.org/abs/1905.13727.
+    Once gradient tensors are aggregated across all workers, this hook applies
+    compression as follows:
+    1) Views the input flattened 1D gradient tensor as a square-shaped tensor M with 0 paddings;
+    2) Decomposes M into two low-rank tensors P and Q,
+    such that M = PQ^T, where Q is initialized from a standard normal distribution;
+    2) Allreduces P;
+    3) Orthogonizes P;
+    4) Compute Q, which is approximately equal to M^TP;
+    5) Allreduces Q;
+    6) Computes M, which is approximately equal to PQ^T.
+    7) Truncates the input tensor to the original length.
```

Comment:
This procedure (as described in the paper) does two matmul+allreduce steps per 'iteration': one left multiplication and one right multiplication. If you wanted, you could take one such step at a time, and alternate between them (for warm start)

Reply:
Thanks for the advice! Are you referring to HalfRankKReducer implementation in your repo? Added a TODO here to try it in a different PR.